Fixed empty table processing on server

ashayshirwadkar opened the pull request:
Current version of Apache Arrow does not support creating
table form empty record batch vector. Therefore, work around
is added for the processing of empty table

ashayshirwadkar committed:
Fixed empty table processing on server

Current version of Apache Arrow does not support creating
table form empty record batch vector. Therefore, work around
is added for the processing of empty table

jlefevre commented on src/progly/query.cc:
```diff
@@ -390,7 +390,7 @@ void worker()
                 int decode_runquery_cls = 0;
                 assert(decode_runquery_cls);
             }
-            nrows_processed += nrows_server_processed;
```
I'm not sure here but as mentioned I think this should be restored to the previous accounting or in some way such that a query that matches say 3/20 rows, should report 3/3 with --use-cls flag and 3/20 without the flag.  Since the cls case should have already filtered out the rows, only the 3 matching rows are actually returned (denominator).  Without cls all 20 rows are returned from server.  The only exception is SELECT*, which does not do any accounting due to fastpath execution and hence the server does not even look at the data/metadata to determine how many rows are present, it just blindly returns the data.   Thanks.

ashayshirwadkar replied:
Done

ashayshirwadkar committed:
Reverted the logic for rows returned from Server

ashayshirwadkar committed:
Changed process arrow function

Arrow being columnr data format processing the columns instead
of rows inside arrow will be beneficial.

jlefevre commented on src/cls/tabular/cls_tabular.cc:
```diff
@@ -1497,7 +1440,7 @@ int exec_query_op(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
                         createFbMeta(meta_builder,
                                      SFT_ARROW,
-                                     (unsigned char *)buffer->ToString().c_str(),
+                                     reinterpret_cast<unsigned char*>(buffer->mutable_data()),
```
Better, thanks.

jlefevre commented on src/cls/tabular/cls_tabular.cc:
```diff
@@ -1525,7 +1468,6 @@ int exec_query_op(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
                         ds_rows_processed = row_nums.size();
```
Looks good, outside the loop now.

jlefevre commented on src/progly/query.cc:
```diff
@@ -466,35 +468,21 @@ void worker()
-                        sky_root root = Tables::getSkyRoot(meta.blob_data,
-                                                           meta.blob_size,
-                                                           SFT_ARROW);
+
```
This is better than previous switch statement, thanks.

jlefevre commented on src/cls/tabular/cls_tabular.cc:
```diff
@@ -1488,7 +1431,7 @@ int exec_query_op(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
 
                         if (ret != 0) {
-                            CLS_ERR("ERROR: processSkyFb %s", errmsg.c_str());
+                            CLS_ERR("ERROR: processArrow %s", errmsg.c_str());
```
good catch.

jlefevre commented on src/cls/tabular/cls_tabular.cc:
```diff
@@ -1423,7 +1367,6 @@ int exec_query_op(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
-
                     case SFT_JSON: {
 
                         sky_root root = \
```
This can be removed from JSON case also, but I will assign to the person working JSON.
